internal/http/client.go

```diff
@@ -381,8 +390,9 @@ func (h *Client) ParseResponse(res *http.Response) (resMsg types.Message, err er
 				if h.conf.CopyResponseHeaders {
 					meta := resMsg.Get(index).Metadata()
 					for k, values := range p.Header {
-						if len(values) > 0 {
-							meta.Set(strings.ToLower(k), values[0])
+						normalisedHeader := strings.ToLower(k)
+						if len(values) > 0 && (!h.metaFilter.IsSet() || h.metaFilter.Match(normalisedHeader)) {
```

Comment:
I'm thinking through the use cases here and I reckon we should maybe flip the logic here so that rather than checking whether the filter is set we have an inclusive-default filter similar to the existing outputs where there's an exclude prefixes, exclude patterns, etc.

The reason for that is the field itself is sort of an indicator as to whether the component without filters defaults to including all metadata or excluding it all. We would see `exclude_*` and would be more likely to infer that the behaviour with zero filters in this case is to copy all headers as metadata.

Reply:
That's very thoughtful, nice catch! It also simplifies the code quite a bit. Thanks @Jeffail 

Just to see if I understood you correctly, there are 2 cases:
- http_client (input, processor, output): The user sets `copy_response_headers: true`, then all headers are included as metadata. The behaviour is the same if they additionally set `metadata_filter: {}`.
- http_server (input): There is no equivalent of `copy_response_headers`, so, after my latest changes, all headers are set as metadata unless there's a `metadata_filter` configured to limit them. Since this alters the existing behaviour, I'm not sure if it's always desirable. Shall I add add an extra flag, like `copy_request_headers`?

Reply:
What we're aiming to end up with is two standards for metadata filtering, one where they are inclusive by default and one where they are exclusive by default. When we have inclusive fields and they aren't set then we shouldn't be adding any metadata fields at all, so the server implementation right now needs this block to be removed:

```go
if len(f.includePrefixes) == 0 && len(f.includePatterns) == 0 {
  return true
}
```

And then it works as I'd expect, with:

```yaml
# Config that doesn't copy any metadata in the sync response as headers:
http_server:
  sync_response:
    status: "200"
    headers:
      Content-Type: application/json

# Config that copies all metadata in the sync response as headers:
http_server:
  sync_response:
    status: "200"
    headers:
      Content-Type: application/json
      metadata:
        include_patterns: [ .* ]
```

However, the client cases are getting super confusing because of `copy_response_headers`. Ideally we wouldn't have that field at all and instead we'd have `include_*` fields working the exact same as `http_server`, as it's basically a more powerful version of it.

Since we _do_ have that field we can either embrace it and add `exclude_*` fields that complement it:

```yaml
# Config that doesn't copy any headers as metadata:
http:
  url: foo
  copy_response_headers: false

# Config that copies all headers as metadata:
http:
  url: foo
  copy_response_headers: true

# Config that copies a subset of headers as metadata:
http:
  url: foo
  copy_response_headers: true
  metadata_filter:
    exclude_patterns: [ foo_.* ]
```

Or, for the nicer more future proof case, we deprecate `copy_response_headers` entirely, and then add `include_*` fields that work in total isolation. If a user sets both `copy_response_headers` to `true` and also configures any `include_` patterns then we log a warning, and potentially also show that as a linting error.

That would give us this:

```yaml
# Config that copies a subset of headers as metadata keys:
http:
  url: foo
  extract_metadata:
    include_patterns: [ foo_.* ]

# Config that copies all headers as metadata keys:
http:
  url: foo
  extract_metadata:
    include_patterns: [ .* ]

# Config that copies all headers as metadata keys using the old deprecated style:
http:
  url: foo
  copy_response_headers: true

# Config that doesn't copy any headers as metadata keys:
http:
  url: foo
```

Reply:
Awesome! I really liked the idea of deprecating `copy_response_headers` and I wanted to get some experience with the linting functionality, so I went full in and implemented your 3rd proposal. For linting to work, I had to do some refactoring in each component which uses `lib/util/http/client/docs.go` because the lint needs to be set on the parent of the `FieldSpecs` and `docs.LintFunc` doesn't seem to provide any means to access the parent `FieldSpec` via the `LintContext`, but please correct me if I'm wrong here.

It might be a good idea to add some extra tests for the `http_client` input and output and the `http` processor that replicate some of the assertions from the `TestHTTPClientReceiveHeadersWithMetadataFiltering` test I added in `internal/http/client_test.go` if you think it's worth it.

One thing to note in the current code is that `include_prefixes: [ "" ]` includes everything, because `strings.HasPrefix()` returns `true` in this case. Somehow, I feel that maybe we should reject this pattern or skip it, but not sure. WDYT?